ProjFS.Mac/PrjFSKextLogDaemon/PrjFSKextLogDaemon.cpp

```diff
@@ -221,12 +239,141 @@ static bool TryFetchAndLogKextHealthData(io_connect_t connection)
+            "\"InvalidateRoot\":\"" + to_string(healthData.totalInvalidateVnodeRoot) + "\","
+        "}");
+        
     }
     else
     {
+        WriteJsonToMessageListener(
+        "{"
+            "\"ErrorMessage\":\"Fetching profiling data from kernel failed\","
+            "\"ret\":\"" + to_string(ret) + "\","
+        "}");
+        
         fprintf(stderr, "fetching profiling data from kernel failed: 0x%x\n", ret);
         return false;
     }
     
     return true;
 }
+
+static void CreatePipeToMessageListener()
+{
+    if (INVALID_SOCKET_FD != s_messageListenerSocket)
+    {
+        // Already connected
+        return;
+    }
+
+    s_messageListenerSocket = socket(PF_UNIX, SOCK_STREAM, 0);
+    if (s_messageListenerSocket < 0)
+    {
+        os_log_with_type(
+            s_kextLogger,
+            OS_LOG_TYPE_DEFAULT,
+            "Failed to create a new socket, path: %s, error: %d",
+            s_messageListenerSocketPath.c_str(),
+            errno);
+        
+        s_messageListenerSocket = INVALID_SOCKET_FD;
+        return;
+    }
+    
+    struct sockaddr_un socket_address;
+    memset(&socket_address, 0, sizeof(struct sockaddr_un));
+    
+    socket_address.sun_family = AF_UNIX;
+    size_t resultLength = strlcpy(socket_address.sun_path, s_messageListenerSocketPath.c_str(), sizeof(socket_address.sun_path));
+    
+    if (resultLength >= sizeof(socket_address.sun_path))
+    {
+        os_log_with_type(
+            s_kextLogger,
+            OS_LOG_TYPE_DEFAULT,
+            "Could not copy socket path: %s, insufficient buffer. resultLength: %lu, sizeof(socket_address.sun_path): %lu",
+            s_messageListenerSocketPath.c_str(),
+            resultLength,
+            sizeof(socket_address.sun_path));
+        
+        goto ClosePipeAndCleanup;
+    }
+    
+    if(0 == connect(s_messageListenerSocket, (struct sockaddr *) &socket_address, sizeof(struct sockaddr_un)))
+    {
+        os_log_with_type(
+            s_kextLogger,
+            OS_LOG_TYPE_DEFAULT,
+            "Connected to message listener on socket '%s'",
+            s_messageListenerSocketPath.c_str());
+        return;
+    }
+    
+    os_log_with_type(
+        s_kextLogger,
+        OS_LOG_TYPE_DEFAULT,
+        "Failed to connect socket, pipeName: %s, error: %d",
+        s_messageListenerSocketPath.c_str(),
+        errno);
+    
+ClosePipeAndCleanup:
+
+    if (INVALID_SOCKET_FD != s_messageListenerSocket)
+    {
+        close(s_messageListenerSocket);
+        s_messageListenerSocket = INVALID_SOCKET_FD;
+    }
+}
+
+static void WriteJsonToMessageListener(const string& jsonMessage)
+{
+    if (INVALID_SOCKET_FD == s_messageListenerSocket)
+    {
+        return;
+    }
+
+    // TODO: Properly version PrjFSKextLogDaemon and other Mac binaries
+    string fullMessage =
+    "{\""
+        "version\":\"0.6.XXX.X\","
+        "\"providerName\":\"Microsoft.Git.GVFS\","
```

Comment:
@jeffhostetler @jrbriggs, let me know your thoughts on sending these messages over using the  "Microsoft.Git.GVFS" provider name.

Reply:
The provider name is an ETW holdover.  I only needed it in the protocol between the mount process and GitService to let me route it to the right GUID in ETW world.  For AppIns, I secretly abbreviate it to “VFS” and combine it with the event name.  That’s why you’re seeing justo events with “vfs.heartbeat” and etc.

It really doesn’t matter in the AI world because I can send everything to the same AI resource id/group (and without having to worry about registering a new (providerName, eventName) pair in some DB somewhere).

Once we drop ETW completely, we can revisit abbreviating them in the mount and kext processes.

Reply:
Alternatively, you could have a kext provider name, but I’d have to add another translation entry to the service.  No big deal either way.

Reply:
> Probably better to have a new provider entry, even if they route to the same AI endpoint.
The usermode and kernel mode pipe message schemas can diverge overtime then.

I think this sounds like something we could do in a follow up work item/PR?  @jeffhostetler would you agree?

Reply:
>I think this sounds like something we could do in a follow up work item/PR? @jeffhostetler would you agree?

The downside to delaying this work is that different versions of PrjFSKextLogDaemon will write the data in different formats, making it more difficult to consume.  Maybe we should get all of this sorted out now?

\+ @jrbriggs for his thoughts

Note to self:  Also need to verify that PrjFSKextLogDaemon is being versioned properly.

Reply:
I’ll be back in the office on Monday and we can talk about stuff then if you want.
We can address the dedicated named pipe/socket, custom proxy, the format 
of the messages, and the changes to the provider and event name that we’ve 
talked about.  (On my iPad here and not sure if I’ve left anything out of this list or not.)

Reply:
I just discussed this offline with @jrbriggs.  I won't hold this PR up on the provider name decision, and @jeffhostetler when you're back in we can decide if we want to update it before the next release.

We were also discussing that leaving the provider name as-is with "vfs.kext.*" events is probably OK too.

Reply:
Closing this thread out, based on our offline discussion we'll leave the provider name as "vfs.kext.*"